Thecosomata (ETH/BTRFLY in CRV)

drahrealm opened the pull request:
## Changes
- Migrated previous Thecosomata WIP to be used for liquidity pool in Curve for ETH-BTRFLY pair

## Tests
- `npx hardhat clean && npx hardhat compile && npx hardhat test`

realkinando committed:
started contract, using Thecosomata as a foundation - doesn't compile

Commit:
Refactor thecosomata to support curve pool for ETH/BTRFLY

Commit:
Transfer out LP token directly to Treasury

Commit:
Add test for lp token balance check

Commit:
Updated test for more use cases

Commit:
Add comment and shorten formula

kphed commented on contracts/thecosomata/ThecosomataETH.sol:
```diff
@@ -118,9 +118,11 @@ contract ThecosomataETH is Ownable {
 
-        require(ethLiquidity > 0 || btrflyLiquidity > 0, "Invalid amounts");
+        // Only complete upkeep only on sufficient amounts
+        require(ethLiquidity > 0 && btrflyLiquidity > 0, "Insufficient amounts");
```
Nice - this was what I was mentioning I'd comment on

kphed commented on test/ThecosomataETH.ts:
```diff
@@ -0,0 +1,215 @@
+    });
+
+    it("Should not perform upkeep on insufficient balance on either token", async () => {
+      // Mint a very small amount of BTRFLY, which would result in 0 amount in ETH
```
This test seems to work as intended (based on the test description) but not because the BTRFLY amount is so small that there is a 0 amount of ETH; `performUpkeep` reverts because there is no ETH left in the treasury so `ethLiquidity` is 0. If the test is run in isolation it actually fails.

Wanted to point this out to confirm that it's in line with what you're expecting.

drahrealm replied:
Good catch, thanks! Will update the unit test

kphed commented on contracts/thecosomata/ThecosomataETH.sol:
```diff
@@ -0,0 +1,145 @@
+        IERC20(_WETH).approve(_CURVEPOOL, 2**256 - 1);
+    }
+
+    function checkUpkeep()
```
Since we're not going to use Chainlink keepers anymore I think we can remove this. Can you confirm @realkinando ?

kphed commented on contracts/thecosomata/ThecosomataETH.sol:
```diff
@@ -0,0 +1,145 @@
+        uint256 priceOracle = ICurveCryptoPool(CURVEPOOL).price_oracle();
+
+        if (isBTRFLY) {
+            return ((amount * priceOracle) / (10**IBTRFLY(BTRFLY).decimals()));
```
Can we set the BTRFLY decimals as an immutable variable in the constructor since it probably won't ever change?

drahrealm replied:
Yeah, that's a good idea

Commit:
Store decimals data on constructor

Commit:
Update price formula and tests

Commit:
Remove console

Commit:
Update comment

kphed commented:
Thanks for the changes @drahrealm. Not necessary, but if we really want to make sure all the #s check out, could set up a Curve pool and make sure their calculations for amounts (when adding liquidity in 1:1 ETH/BTRFLY ratio) matches ours.

Let's hit up @realkinando and wrap this up.

Commit:
Add support for slippage when adding liquidity

Commit:
Add tests for slippage related logic

Commit:
Refactor upKeep flow

- `performUpKeep` now requires a parameter specifying the minimum amount of LP token received (for slippage handling).
- Use `getMinimumLPAmount` instead of now removed `checkUpkeep` as an indicator whether to perform upKeep or not.
- Gas optimizations

drahrealm commented:
@realkinando I did a major refactor to the overall flow of the contract, mostly related to the findings in our latest audit regarding:
- Added capability to prevent MEV attack by requiring `performUpkeep` to accept a parameter specifying the minimum amount of LP token received
- Removed `checkUpkeep` and instead added `getMinimumLPAmount` which is used by the off-chain script/process to decide whether to perform upkeep or not, as well to pass over its value when calling `performUpkeep`
- Fix security issue by using SafeERC20 standard for token transfers
- Gas optimizations

Commit:
Add separate check for min lpToken specified

Commit:
Add withdrawal event

Commit:
Use type max val for clarity

kphed commented on contracts/thecosomata/ThecosomataETH.sol:
```diff
@@ -0,0 +1,190 @@
+    function token() external view returns (address);
+}
+
+contract ThecosomataETH is Ownable {
```
A Code4rena comment made me realize that we should use the AccessControl instead of Ownable, because we should have a new role that is allowed to call `perofrmUpkeep` but not `withdraw`

kphed replied:
Explaining the above: we want to do this because the owner/admin should be the multisig and should be the only one able to withdraw tokens. We should limit our own off-chain process or keeper contracts to calling only `performUpkeep`

drahrealm replied:
That makes sense!

Commit:
Use access control instead of ownable for more specific access management

Commit:
Add extra sanity check for price oracle

Commit:
Update comment

Commit:
Use constant instead of immutable for role hashes

Commit:
Remove obsolete comment

kphed commented on contracts/thecosomata/ThecosomataETH.sol:
```diff
@@ -0,0 +1,219 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.0;
```
Let's pin this

kphed commented on contracts/thecosomata/ThecosomataETH.sol:
```diff
@@ -0,0 +1,219 @@
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+interface IBTRFLY is IERC20 {
```
Just looked at IERC20.sol and surprised that it doesn't have the burn and decimals methods (not to mention others).

Can we rename this `IERC20Extended` since we're using it once for eth?

Commit:
Update semantics

Commit:
Revert to on-chain calculation for min. lp token expected

Commit:
Re-add ability to apply slippage

realkinando reviewed (approved):
has been thoroughly reviewed via code4rena audit and consultation from MEV experts
